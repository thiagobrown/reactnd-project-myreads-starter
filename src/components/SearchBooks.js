import React, { Component } from 'react'
import { Link } from 'react-router-dom'
import { Debounce } from 'react-throttle';
import PropTypes from 'prop-types'

import * as BooksAPI from './../utils/BooksAPI'

import Book from './Book'
import Loader from './Loader'

class SearchBooks extends Component {

    state = {
        searchBooks: [],
        isLoading: false
    }

    static propTypes = {
        books: PropTypes.array,
        onUpdateBook: PropTypes.func.isRequired
    }

    onSearch = (query) => {
        if (query && query.length) {
            this.setState({ isLoading: true });
            BooksAPI.search(query, 20).then((books) => {
                this.setState({
                    searchBooks: this.onChangeShelfSearchBooks(books),
                    isLoading: false
                });
            }).catch(error => this.setState({ searchBooks: [] }));
        } else {
            this.setState({
                searchBooks: [],
                isLoading: false
            });
        }
    }

    onChangeShelfSearchBooks = (searchBooks) => {
        if (searchBooks && Array.isArray(searchBooks)) {
            searchBooks.forEach(searchBook => {
                searchBook.shelf = "none";
                this.props.books.forEach(book => {
                    if (searchBook.id === book.id) {
                        searchBook.shelf = book.shelf;
                    }
                });
            });

            return searchBooks;
        }

        return [];
    }

    render() {
        return (
            <div className="search-books">
                <div className="search-books-bar">
                    <Link className="close-search" to="/" />
                    <div className="search-books-input-wrapper">
                        <Debounce time="500" handler="onChange">
                            <input
                                type="text"
                                placeholder="Search by title or author"
                                onChange={event => this.onSearch(event.target.value)} />
                        </Debounce>
                    </div>
                </div>
                <div className="search-books-results">
                    {
                        this.state.isLoading
                            ? (<Loader />)
                            : (
                                <ol className="books-grid">
                                    {this.state.searchBooks.map(b => (
                                        <Book key={b.id} book={b} onUpdateBook={this.props.onUpdateBook} />
                                    ))}
                                </ol>
                            )
                    }
                </div>
            </div>
        )
    }
}

export default SearchBooks;
