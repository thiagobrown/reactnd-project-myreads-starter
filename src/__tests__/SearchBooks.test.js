import React from 'react'
import ReactDOM from 'react-dom'
import { Debounce } from 'react-throttle';
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json';

import SearchBooks from './../components/SearchBooks'
import Loader from './../components/Loader'

describe('Testing SearchBooks Component', () => {
  let props = {
    books: [{
      id: 1,
      title: 'Book One',
      authors: ['Authors One'],
      imageLinks: { thumbnail: 'url/book/one' },
      shelf: 'read'
    }, {
      id: 2,
      title: 'Book Two',
      authors: ['Authors Two'],
      shelf: 'wantToRead'
    }, {
      id: 3,
      title: 'Book Three',
      imageLinks: { thumbnail: 'url/book/three' },
      shelf: 'currentlyReading'
    }],
    onUpdateBook: jest.fn()
  }

  const wrapper = shallow(<SearchBooks {...props} />);

  it('render correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('check render when state changes', () => {
    wrapper.setState({
      books: [
        {
          id: 1,
          title: 'Book One',
          authors: ['Authors One'],
          imageLinks: { thumbnail: 'url/book/one' },
          shelf: 'read'
        }, {
          id: 2,
          title: 'Book Two',
          authors: ['Authors Two'],
          shelf: 'wantToRead'
        }, {
          id: 3,
          title: 'Book Three',
          imageLinks: { thumbnail: 'url/book/three' },
          shelf: 'currentlyReading'
        }]
    });

    expect(wrapper).toMatchSnapshot();
  });

  it('check component if exists Debouce', () => {
    expect(wrapper.exists(<Debounce />)).toEqual(true);
  });

  it('Debouce props.time = 500', () => {
    expect(wrapper.find(Debounce).prop('time')).toBe('500')
  });

  it('Debouce props.handler = onChange', () => {
    expect(wrapper.find(Debounce).prop('handler')).toBe('onChange')
  });

  it('input changes the text', () => {
    wrapper.find('input').prop('onChange')({ target: { value: 'React' } });
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('check component <Loader/> when state.loading change to true', () => {
    wrapper.setState({ isLoading: true });
    expect(wrapper.containsAllMatchingElements([<Loader />])).toEqual(true);
  });

  it('check component <Loader/> when state.loading change to false', () => {
    wrapper.setState({ isLoading: false });
    expect(wrapper.containsAllMatchingElements([<Loader />])).toEqual(false);
  });
});