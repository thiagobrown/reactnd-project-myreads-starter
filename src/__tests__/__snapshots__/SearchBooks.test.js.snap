exports[`Testing SearchBooks Component check render when state changes 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchBooks
    books={
        Array [
            Object {
              "authors": Array [
                "Authors One",
              ],
              "id": 1,
              "imageLinks": Object {
                "thumbnail": "url/book/one",
              },
              "shelf": "read",
              "title": "Book One",
            },
            Object {
              "authors": Array [
                "Authors Two",
              ],
              "id": 2,
              "shelf": "wantToRead",
              "title": "Book Two",
            },
            Object {
              "id": 3,
              "imageLinks": Object {
                "thumbnail": "url/book/three",
              },
              "shelf": "currentlyReading",
              "title": "Book Three",
            },
          ]
    }
    onUpdateBook={[Function]} />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="search-books-bar">
          <Link
                    className="close-search"
                    replace={false}
                    to="/" />
          <div
                    className="search-books-input-wrapper">
                    <Debounce
                              handler="onChange"
                              handlers={Array []}
                              time="500">
                              <input
                                        onChange={[Function]}
                                        placeholder="Search by title or author"
                                        type="text" />
                    </Debounce>
          </div>
</div>,
        <div
          className="search-books-results">
          <ol
                    className="books-grid">
                    
          </ol>
</div>,
      ],
      "className": "search-books",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Link
              className="close-search"
              replace={false}
              to="/" />,
            <div
              className="search-books-input-wrapper">
              <Debounce
                            handler="onChange"
                            handlers={Array []}
                            time="500">
                            <input
                                          onChange={[Function]}
                                          placeholder="Search by title or author"
                                          type="text" />
              </Debounce>
</div>,
          ],
          "className": "search-books-bar",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "className": "close-search",
              "replace": false,
              "to": "/",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Debounce
                handler="onChange"
                handlers={Array []}
                time="500">
                <input
                                onChange={[Function]}
                                placeholder="Search by title or author"
                                type="text" />
</Debounce>,
              "className": "search-books-input-wrapper",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <input
                  onChange={[Function]}
                  placeholder="Search by title or author"
                  type="text" />,
                "handler": "onChange",
                "handlers": Array [],
                "time": "500",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Search by title or author",
                  "type": "text",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ol
            className="books-grid">
            
</ol>,
          "className": "search-books-results",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "books-grid",
          },
          "ref": null,
          "rendered": Array [],
          "type": "ol",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="search-books-bar">
            <Link
                        className="close-search"
                        replace={false}
                        to="/" />
            <div
                        className="search-books-input-wrapper">
                        <Debounce
                                    handler="onChange"
                                    handlers={Array []}
                                    time="500">
                                    <input
                                                onChange={[Function]}
                                                placeholder="Search by title or author"
                                                type="text" />
                        </Debounce>
            </div>
</div>,
          <div
            className="search-books-results">
            <ol
                        className="books-grid">
                        
            </ol>
</div>,
        ],
        "className": "search-books",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                className="close-search"
                replace={false}
                to="/" />,
              <div
                className="search-books-input-wrapper">
                <Debounce
                                handler="onChange"
                                handlers={Array []}
                                time="500">
                                <input
                                                onChange={[Function]}
                                                placeholder="Search by title or author"
                                                type="text" />
                </Debounce>
</div>,
            ],
            "className": "search-books-bar",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "className": "close-search",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Debounce
                  handler="onChange"
                  handlers={Array []}
                  time="500">
                  <input
                                    onChange={[Function]}
                                    placeholder="Search by title or author"
                                    type="text" />
</Debounce>,
                "className": "search-books-input-wrapper",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <input
                    onChange={[Function]}
                    placeholder="Search by title or author"
                    type="text" />,
                  "handler": "onChange",
                  "handlers": Array [],
                  "time": "500",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "Search by title or author",
                    "type": "text",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ol
              className="books-grid">
              
</ol>,
            "className": "search-books-results",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "books-grid",
            },
            "ref": null,
            "rendered": Array [],
            "type": "ol",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;

exports[`Testing SearchBooks Component input changes the text 1`] = `
<div
  className="search-books">
  <div
    className="search-books-bar">
    <Link
      className="close-search"
      replace={false}
      to="/" />
    <div
      className="search-books-input-wrapper">
      <Debounce
        handler="onChange"
        handlers={Array []}
        time="500">
        <input
          onChange={[Function]}
          placeholder="Search by title or author"
          type="text" />
      </Debounce>
    </div>
  </div>
  <div
    className="search-books-results">
    <ol
      className="books-grid" />
  </div>
</div>
`;

exports[`Testing SearchBooks Component render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SearchBooks
    books={
        Array [
            Object {
              "authors": Array [
                "Authors One",
              ],
              "id": 1,
              "imageLinks": Object {
                "thumbnail": "url/book/one",
              },
              "shelf": "read",
              "title": "Book One",
            },
            Object {
              "authors": Array [
                "Authors Two",
              ],
              "id": 2,
              "shelf": "wantToRead",
              "title": "Book Two",
            },
            Object {
              "id": 3,
              "imageLinks": Object {
                "thumbnail": "url/book/three",
              },
              "shelf": "currentlyReading",
              "title": "Book Three",
            },
          ]
    }
    onUpdateBook={[Function]} />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="search-books-bar">
          <Link
                    className="close-search"
                    replace={false}
                    to="/" />
          <div
                    className="search-books-input-wrapper">
                    <Debounce
                              handler="onChange"
                              handlers={Array []}
                              time="500">
                              <input
                                        onChange={[Function]}
                                        placeholder="Search by title or author"
                                        type="text" />
                    </Debounce>
          </div>
</div>,
        <div
          className="search-books-results">
          <ol
                    className="books-grid">
                    
          </ol>
</div>,
      ],
      "className": "search-books",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Link
              className="close-search"
              replace={false}
              to="/" />,
            <div
              className="search-books-input-wrapper">
              <Debounce
                            handler="onChange"
                            handlers={Array []}
                            time="500">
                            <input
                                          onChange={[Function]}
                                          placeholder="Search by title or author"
                                          type="text" />
              </Debounce>
</div>,
          ],
          "className": "search-books-bar",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "className": "close-search",
              "replace": false,
              "to": "/",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <Debounce
                handler="onChange"
                handlers={Array []}
                time="500">
                <input
                                onChange={[Function]}
                                placeholder="Search by title or author"
                                type="text" />
</Debounce>,
              "className": "search-books-input-wrapper",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <input
                  onChange={[Function]}
                  placeholder="Search by title or author"
                  type="text" />,
                "handler": "onChange",
                "handlers": Array [],
                "time": "500",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "onChange": [Function],
                  "placeholder": "Search by title or author",
                  "type": "text",
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              "type": [Function],
            },
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <ol
            className="books-grid">
            
</ol>,
          "className": "search-books-results",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "books-grid",
          },
          "ref": null,
          "rendered": Array [],
          "type": "ol",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="search-books-bar">
            <Link
                        className="close-search"
                        replace={false}
                        to="/" />
            <div
                        className="search-books-input-wrapper">
                        <Debounce
                                    handler="onChange"
                                    handlers={Array []}
                                    time="500">
                                    <input
                                                onChange={[Function]}
                                                placeholder="Search by title or author"
                                                type="text" />
                        </Debounce>
            </div>
</div>,
          <div
            className="search-books-results">
            <ol
                        className="books-grid">
                        
            </ol>
</div>,
        ],
        "className": "search-books",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Link
                className="close-search"
                replace={false}
                to="/" />,
              <div
                className="search-books-input-wrapper">
                <Debounce
                                handler="onChange"
                                handlers={Array []}
                                time="500">
                                <input
                                                onChange={[Function]}
                                                placeholder="Search by title or author"
                                                type="text" />
                </Debounce>
</div>,
            ],
            "className": "search-books-bar",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "className": "close-search",
                "replace": false,
                "to": "/",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <Debounce
                  handler="onChange"
                  handlers={Array []}
                  time="500">
                  <input
                                    onChange={[Function]}
                                    placeholder="Search by title or author"
                                    type="text" />
</Debounce>,
                "className": "search-books-input-wrapper",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "class",
                "props": Object {
                  "children": <input
                    onChange={[Function]}
                    placeholder="Search by title or author"
                    type="text" />,
                  "handler": "onChange",
                  "handlers": Array [],
                  "time": "500",
                },
                "ref": null,
                "rendered": Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "host",
                  "props": Object {
                    "onChange": [Function],
                    "placeholder": "Search by title or author",
                    "type": "text",
                  },
                  "ref": null,
                  "rendered": null,
                  "type": "input",
                },
                "type": [Function],
              },
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <ol
              className="books-grid">
              
</ol>,
            "className": "search-books-results",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [],
              "className": "books-grid",
            },
            "ref": null,
            "rendered": Array [],
            "type": "ol",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
